### PR TITLE
Distributed matrix-matrix multiplication

### PR DESCRIPTION
**Pull request by rahulghangas**
A general distributed matrix-matrix multiplication procedure.

- [x] Works for multiple different distributions (Block, Cyclic, BlockCyclic)
- [x] Works if input matrices have different `targetLocales` or entirely different distributions
- [x] Multilocale testing - Correctness
- [x] Multilocale testing - Perf (is this needed if PRK DGEMM already exists)
- [ ] Perf test against MPI reference

### CODE REVIEW

**Comment by rahulghangas** [comment]
We need an overload of `dot` or `_matmatmult` (preferably the second one) that takes in the output array as a parameter

**Comment by rahulghangas** [comment]
@ben-albrecht @dgarvit @e-kayrakli @LouisJenkinsCS

**Comment by rahulghangas** [comment]
Just realized there is a lot of stuff wrong with this PR. Converting back to draft

File: modules/packages/LinearAlgebra.chpl [hunk 1]
```diff
@@ -773,6 +773,97 @@ proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
...
+    halt("Array dimensions don't match.\n Trying to multiply arrays of dimensions ", 
+         A.shape, " ", B.shape);
+  }
+
+  const commonDim = Adom.dim(1);
+  if window < 1 {
+    window = commonDim.size;
+  }
+
+  ref Bref = B.reindex(commonDim, Bdom.dim(1));
```
Comment (dgarvit) [inline modules/packages/LinearAlgebra.chpl:825]: I don't think this is used anywhere.

File: modules/packages/LinearAlgebra.chpl [hunk 2]
```diff
@@ -773,6 +773,97 @@ proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
...
+}
+
+
+pragma "no doc"
+/* Distributed matrix-matrix multiplication */
+proc _matmatMult(A : [?Adom] ?eltType, B : [?Bdom] eltType, in window : int = -1)
+  where isDistributed(A) || isDistributed(B)
+{
+  ref targetLocales = A.targetLocales();
+
```
Comment (dgarvit) [inline modules/packages/LinearAlgebra.chpl:814]: I think we should add a rank check.
```
if Adom.rank != 2 || Bdom.rank != 2 then
    compilerError("Ranks are not 2");
```

File: modules/packages/LinearAlgebra.chpl [hunk 3]
```diff
@@ -773,6 +773,97 @@ proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
...
+  if A.shape(1) != B.shape(0) {
+    halt("Array dimensions don't match.\n Trying to multiply arrays of dimensions ",
```
Comment (dgarvit) [inline modules/packages/LinearAlgebra.chpl:816]: For consistency, I'd suggest using the standard error message used throughout the library for this situation: "Mismatched shape in matrix-matrix multiplication"

File: modules/packages/LinearAlgebra.chpl (hunk 1 above)
Comment (rahulghangas) [reply modules/packages/LinearAlgebra.chpl:825]: Yep, it rolled over from the older version of the implementation

**Comment by e-kayrakli** [comment]
I think LinearAlgebra tests don't use distributed arrays and as noted in the OP we need those tests to be modified to use distributed arrays (maybe in combination with non-distributed ones) for correctness.

I also think that we need a performance test that uses distributed arrays with LinearAlgebra.dot even if the implementation is the same as PRK-DGEMM (it is not guaranteed to stay that way)

**Comment by rahulghangas** [comment]
Actually, I realized that it has already changed from PRK-DGEMM when I added the new block iterator and single initialization of local arrays

**Comment by rahulghangas** [comment]
Will it be okay to copy the test from PRK-DGEMM as is and refactor it for distributed dot?

File: test/library/packages/LinearAlgebra/performance/dot-perf.chpl [hunk 4]
```diff
@@ -19,7 +19,9 @@ config type eltType = real;
 const nbytes = numBytes(eltType);
 
 proc main() {
-  const dom = {1..n, 1..n};
+  const dom = {1..n, 1..n} dmapped if useBlockDist then
+                      new dmap(new Block(boundingBox={1..n, 1..n})) else
```
Comment (e-kayrakli) [inline test/library/packages/LinearAlgebra/performance/dot-perf.chpl:23]: use BlockDist

**Comment by rahulghangas** [comment]
Fixed. I'm running the test now to make sure it compiles and runs

**Comment by rahulghangas** [comment]
Compiles and runs

**Comment by e-kayrakli** [comment]
I'll re-test and review this this week. @rahulghangas -- sorry for the delay with the release and everything. If you could summarize what this PR is about in the PR message, that'd be great. Remember that they go into git history directly, so details can be helpful. Thanks!

**Comment by e-kayrakli** [comment]
Based on silence (and conflicting files), I am closing this PR.

@rahulghangas -- Please feel free to reopen when you have the time to work on this.

**Pull request closed without merging**